### ISSUE TITLE
Pass fitted or unfitted objects to Classification Visualizers - how to avoid fitting X_train twice?

### ISSUE THREAD

**Issue by edesz** [DistrictDataLabs/yellowbrick#623]
For yellowbrick `visualize()` - from the __Classification Visualizers__ section:
- I am trying to avoid fitting a classification model twice to training data. Sould we pass fitted or unfitted model/gridsearch objects to the `visualize` method?
- my question is about when to call `visualizer.fit()`
  - before or after performing a `GridSearchCV`?

Here is a minimum example to recreate my question
- I have a basic [`scikit-learn` pipeline that performs a simple grid search](https://stackoverflow.com/q/37021338/4057186)

Get [Occupancy dataset](https://archive.ics.uci.edu/ml/datasets/Occupancy%2BDetection%2B)
```
$ wget https://raw.githubusercontent.com/LuisM78/Occupancy-detection-data/master/datatest2.txt
```

Code
```
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from yellowbrick.classifier import DiscriminationThreshold
import pandas as pd
from sklearn.grid_search import GridSearchCV
from sklearn.pipeline import Pipeline


def my_plotting_function(f, X_train, y_train):
    visualizer = DiscriminationThreshold(f)

    # visualizer.fit(X_train, y_train)
    visualizer.poof()

    return visualizer


# Load the classification data set
data = pd.read_csv('datatest2.txt')

# Specify the features of interest and the classes of the target
features = ["Temperature", "HumidityRatio", "Light", "CO2", "Humidity"]
classes = [0, 1]

# Extract the numpy arrays from the data frame
X = data[features].as_matrix()
y = data.Occupancy.as_matrix()

# Create the train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)


# Instantiate the classification model and visualizer
mypipe = Pipeline(steps=[('forest', RandomForestClassifier())])
mydict = {'forest__n_estimators': [10,11]}

gs = GridSearchCV(mypipe, mydict, scoring='roc_auc')
gs.fit(X_train, y_train)
print(gs.best_score_)

vz = my_plotting_function(gs, X_train, y_train)
print(vz.best_score_)
```

In the [__Class Balance__ docs](http://www.scikit-yb.org/en/latest/api/classifier/class_balance.html#module-yellowbrick.classifier.class_balance), it says

> model: estimator

    Scikit-Learn estimator object. Should be an instance of a classifier,
    else __init__() will raise an exception.

though I couldn't find if it requires fitted or unfitted estimators(???)

__Questions__

__Case A__
Do I pass the fitted grid search object to the [Classification Visualizers](http://www.scikit-yb.org/en/latest/api/classifier/index.html#classification-visualizers) (I would prefer to do this, if possible)?

    gs = GridSearchCV(mypipe, mydict, scoring='roc_auc')
    gs.fit(X_train, y_train)
    visualizer = DiscriminationThreshold(f)

    # visualizer.fit(X_train, y_train)
    visualizer.poof()

__Case B__
Or must I pass only the unfitted objects to the `visualizer`?

    gs = GridSearchCV(mypipe, mydict, scoring='roc_auc')
    # gs.fit(X_train, y_train)
    visualizer = DiscriminationThreshold(f)

    visualizer.fit(X_train, y_train)
    visualizer.poof()

For the case of `DiscriminationThreshold` in the above example

 - if `.fit()` is removed (__Case A__), then I get a blank plot
 - if `.fit()` is not removed (__Case B__), then the plot is produced (but I don't know if this is fitting the classification model twice to the training data (`X_train`, `y_train`).

**Comment by edesz** [comment]
From trial and error:

Based on my question, I think it is acceptable to use `gs.fit followed by `visualizer.fit` (i.e. it is acceptable to pass a fitted grid search object to the Yellowbrick Classification Visualizers) - I am getting the same output for `best_score_` after each of these so I don't think it is fitting the data twice.

Feel free to correct me if I'm wrong though.

**Comment by bbengfort** [comment]
@edesz sorry it's taken us a while to get back to you - it's been a very busy month for us! 

To your question, I think you got lucky* with the `DiscriminationThreshold` visualizer because it actually does a cross-validation for each threshold, [cloning the estimator](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/classifier/threshold.py#L296) to ensure that each threshold/split pair is correctly scored. Unfortunately, the base `ClassificationVisualizer` does call [fit on the estimator again](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/classifier/base.py#L104). 

This is a known issue (#297) - we wanted to use `sklearn.utils.check_is_fitted` but unfortunately that method requires us to know details about the estimator being passed, which we do not. 

In the meantime, could I suggest the following? 

```python
mypipe = Pipeline(steps=[('forest', RandomForestClassifier())])
mydict = {'forest__n_estimators': [10,11]}

gs = GridSearchCV(mypipe, mydict, scoring='roc_auc', cv=3, refit=False)

# Note that train test splits are not necessary since gs is doing cross-validation
gs.fit(X, y)

estimator = mypipe.set_params(gs.best_params_) 
viz = DiscriminationThreshold(estimator) 

# Note that train test splits are not necessary since viz is doing cross-validation
viz.fit(X, y)
viz.poof()
```

The `gs.best_params_` is available when `refit=False` because you're specifying a single metric, `roc_auc`; this will ensure that you're limiting the number of estimator fits that occurs overall in your pipeline, and getting the best discrimination threshold analysis. Note that if you want to use a multi-metric grid search, you'll have to specify `refit=True` anyway, at which point you could use `gs.best_estimator_` and pass that to the classification visualizer. 

We're also working on visual pipelines #231 which may also help eliminate estimator refits. We'd love some help with these tasks if you were interested in submitting a PR! 

*Lucky here being relative, because it is possible that the `GridSearch` was refit multiple times ...

**Comment by edesz** [comment]
@bbengfort My apologies for the late reply. I;ve been a little busy recently. Ok, if I have some time I'll check out the pipelines link you posted.

For your suggested answer, it definitely works. Thanks for posting this....it seems like a sensible approach to me. Thanks again. Fine to close this thread....I should have closed it sooner.

**Comment by bbengfort** [comment]
@edesz no worries at all; glad to hear the suggested answer works! Please let us know if you have any other questions.
